server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -444,7 +446,7 @@ static void collectSearchShards(IndicesOptions indicesOptions, String preference
             final String[] indices = entry.getValue().indices();
             ClusterSearchShardsRequest searchShardsRequest = new ClusterSearchShardsRequest(indices)
                 .indicesOptions(indicesOptions).local(true).preference(preference).routing(routing);
-            clusterClient.admin().cluster().searchShards(searchShardsRequest,
+            new OriginSettingClient(clusterClient, GetTaskAction.TASKS_ORIGIN).admin().cluster().searchShards(searchShardsRequest,
```

Comment:
@tvernum I'd like to have your view on this before I proceed to complete the PR. Thanks!

In https://github.com/elastic/elasticsearch/issues/58476#issuecomment-668549239 I suggested we treat the searchShards action an implementation detail and perform it with a system user because this action is not needed by default (CCS roundtrip minimized). I think that's the right thing to do and it is exactly what I did here. But the situation is a bit more involved than my original analysis and deserves another look.

Unlike other internal actions, e.g. indexing an API document using `_xpack_security` user when handling CreateApiKey API, this action is performed before the main search action. That is, at this point, we don't know whether the user has privileges for the subsequent search yet. So the workflow is as the follows:

```
perform remote searchShards using system user
perform remote search with the actual user <--- this can fail due to missing privilege
```

If the search request is rejected at the second line due to missing privilege, we have to issue 2 requests (searchShards and search) to the remote cluster to find it out. Compared to the existing logic, where searchShards is executed with the actual user, only 1 request is needed for failure scenario. Plus this one failure is quicker because the searchShards request is simply rejected on the remote cluster without running. 

Ideally we should not need to run the first line if the second line is bound to fail, i.e. something like:

```
if (user has read privileges for target indices on the remote cluster) {
    perform search shard using system user
    perform search with the actual user
} else {
    reject
}
```

The above is basically asking for cross cluster support for has privileges  (#67798).

Overall, I think it is fine to go with current changes and migrate to "has privileges" when it supports remote cluster because it is better to get the deprecation out earlier. Yes it will take a bit longer and waste some resources on the remote cluster for the failure scenario due to an additional roundtrip. But I feel it's acceptable given it does not affect the success scenario. So unless we can prioritize #67798, I suggest we go with this approach first.

Reply:
> I suggested we treat the searchShards action an implementation detail and perform it with a system user 

I'm still not convinced by this argument. Everything is an implementation detail, that doesn't mean we run it as a system user.
There is a real user here, why shouldn't the user be able to perform `searchShards` ?  
There are only 2 reasons to run something as a system user:

1. if we explicitly don't want the user to have access.
2. we _need_ the action to succeed even when a user doesn't have access. 

Do either of those apply here?

Reply:
> 1. if we explicitly don't want the user to have access.

Yes because local cluster search also needs to perform `searchShards` and it is done by [a direct call to the service](https://github.com/elastic/elasticsearch/blob/7.13/server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java#L625). Though it is not exactly the same as run-with-the-system-user, it also bypasses the user security context. If we draw a parallel when searching remote cluster, it is arguable that `searchShards` should be executed with system context. _This is one of the main reason that I consider it as an implementation detail. The other main reason is that this action is performed differently depending on the request setup_ (more on this below).

> 2. we need the action to succeed even when a user doesn't have access.

Ideally No because local cluster search does not execute this action if the user cannot `read` target indices. When search is local, `searchShards` is gated behind the `read` privilege so that it won't run if `read` check fails, i.e. the workflow is:
```
// ===== On local
if can `read` index
    searchShards with direct call to service
    search
else
    error
```

When the search is remote, there is no local check and everything must be checked on the remote cluster. If roundtrip is minimized (default behaviour), things behave very similar to local cluster search and `searchShards` action also doesn't run if the user cannot `read` target indices. Only this time the `read` check if on the remote cluster, i.e. the above workflow is simply moved to the remote cluster.
```
// ===== On local
send ccs search request to remote
// ===== On remote
if can `read` index
    searchShards with direct call to service
    search
else
    error
```

**The complication happens when roundtrip is not minimized**, which is the main issue of this PR. In this setup, the local cluster needs to collect remote shard information first to combine remote and local shards for better accuracy. But since its a remote cluster, we cannot no longer directly call the service like when it's local. Instead, we need to send a request for it and this is where the user context vs. system context problem comes in.  The workflow is changed as follows:
```
// ===== On local
send searchShards request to remote (THE QUESTION: with user or system context?)
// ===== On remote
if can searchShards
    searchShards
else
    error
// ===== On local
send shard search request to remote
// ===== On remote
if can `read` index
    shard search
else
    error
```

Today, the above `searchShards` request is sent with the user context and therefore the user must be able to execute`indices:admin/shards/search_shards` (covered by `read_cross_cluster`). But the similar privilege is never required when search is local. It is really an artifact due to CCS trying to simulate the local cluster search. But if the workflow of local cluster search is indeed the reference to follow, to faithfully simulate the workflow, one should also gate the remote `searchShards` with a remote `read` check and not worry about `searchShards` itself. This is where I see a remote `has_privileges` call for `read` access comes in. But since we don't have it, the other option is to just run `searchShards` with system context. If the user doesn't have read access, the actual search will still be rejected at shard level. 

I hope this makes sense and I am happy to have a high bandwidth conversation on this topic if needed. Thanks!

Reply:
OK, all of that makes sense, but why can't we just add `indices:admin/shards/search_shards` to the `read` privilege?

Then when the roundtrip is not minimized, it would achieve your goal of only running the action if it is authorized.

I'm OK with using a system context instead, but that's always risky - we have to be 100% sure we never do anything in that context that would expose data incorrectly - so if changing `read` is acceptable, it seems like a simpler option.

Reply:
Allow `read` to cover `searchShards` is an option. But there are 2 concerns. First is the obvious question of whether it fits well for what `read` implies. Second and probably more importantly is that a cluster with this change would not be able to CCS another cluster that is without this change. This is also why I didn't suggest to add a new system user with a very restrictive role for `searchShards`. Technically this can be achieved with an API key, but I am not sure whether it is worth the complexity?

Reply:
Just realised that the API key feature can be disabled in remote cluster which means there is no guarantee that an API key would work in all cases. 

> that's always risky - we have to be 100% sure we never do anything in that context that would expose data incorrectly

The final response is still gated by the `read` operation so it is fine in this case. But I agree this is a valid general concern. For example, we have to make sure future changes won't accidentally put more operations inside the system context. Another thing is that the remote cluster will have to do more work to just reject the search call. An extra roundtrip and handling of the `searchShards` request are needed. `searchShards` is a `masterOpertaion` which could mean another network hop to the master node. Overall it is more expensive compared to the existing implementation where the rejection happens in the authorization layer on the first node.

Overall, if we are after an option that can work in all cases, I stilll think it is likely either the system user approach or it needs a cross cluster hasPrivileges call.

Reply:
> Second and probably more importantly is that a cluster with this change would not be able to CCS another cluster that is without this change.

Maybe I'm missing something, but it seems reasonable to me that the permissions you need on the remote cluster are dependent on the version of the remote cluster. That is, prior to 7.14 you might need `read_cross_cluster`, but from 7.14 onwards you don't.

I think the system user proposal is worse in this regard. You propose to deprecate `read_cross_cluster` which is defined on the remote cluster, and replace it with something (context switch to system user) that runs on the local cluster.
If your local cluster is 7.13 and your remote is 7.15 then the remote will generate deprecation warnings, but you cannot  fix them because the cluster running the searching doesn't support the replacement.

Reply:
> If your local cluster is 7.13 and your remote is 7.15 then the remote will generate deprecation warnings, but you cannot fix them because the cluster that is actually performing the search doesn't support the replacement.

This is a great point. Thanks! It makes sense that the deprecation warning and proposed fix should be applied to the same cluster. 

From this perspective, changing `read` privilege is the only viable proposal left. Are we happy to allow `read` to cover `searchShards`? It returns a bunch of information about the nodes, indices and shards, e.g.:
```json
{
  "nodes": {
    "DzTkDdy3Qqm40Fl74kIrrA": {
      "name": "node",
      "ephemeral_id": "mBksmK56QZqO6OWRKuTkVA",
      "transport_address": "127.0.0.1:9300",
      "attributes": {
        "ml.machine_memory": "34359738368",
        "xpack.installed": "true",
        "ml.max_open_jobs": "512",
        "ml.max_jvm_size": "1073741824"
      },
      "roles": [
        "data",
        "data_cold",
        "data_content",
        "data_frozen",
        "data_hot",
        "data_warm",
        "ingest",
        "master",
        "ml",
        "remote_cluster_client",
        "transform"
      ]
    }
  },
  "indices": {
    "test": : {
      "aliases": [
        "alias1"
      ]
    }
  },
  "shards": [
    [
      {
        "state": "STARTED",
        "primary": true,
        "node": "DzTkDdy3Qqm40Fl74kIrrA",
        "relocating_node": null,
        "shard": 0,
        "index": "test",
        "allocation_id": {
          "id": "4L0GCCy-SweaH70Nq4NObA"
        }
      }
    ]
  ]
}
```
It could be quite a stretch to justify all above to be covered the `read` privilege, especially since:
1. It reveals aliases. If the request targets a filtered alias, it gives back the filter query.
2. It gives back a bunch of node information including IP addresses.

But if changing "read" is not an option, any other "remote-cluster-only" solutions would likely step into the territory of "rethinking the cross cluster security model" which is too big to handle right now?